Create Input, Fieldset, and Field components for terra-form

JakeLaCombe opened the pull request:
### Summary
Our last forms pull request was massive 😮 .

Let's make things a littler easier, by having small plates at a time!

For this review, we will start with the standard Field, Fieldset, and Input elements. Those will be the base for other individual terra components.

### Additional Details
Once this review is done, I'll start working on the Control component next as that will have more complexity, followed by TextField, Numberfield, etc.

bjankord commented on packages/terra-form/src/components/Field.scss:
```diff
@@ -0,0 +1,55 @@
+@import '~terra-mixins';
+@import '../_variables.scss';
```
When importing a partial .scss file, you can omit the underscore prefix.
e.g. `@import '../variables.scss';`

JakeLaCombe replied:
Fixed it here - https://github.com/cerner/terra-core/pull/352/commits/594716abdc84473dfcf3ced97dc9fb5897d5666c

bjankord commented on packages/terra-form/src/components/Input.scss:
```diff
@@ -0,0 +1,18 @@
+@import '~terra-mixins';
+@import '../_variables.scss';
```
When importing a partial .scss file, you can omit the underscore prefix.
e.g. `@import '../variables.scss';`

JakeLaCombe replied:
Fixed it here - https://github.com/cerner/terra-core/pull/352/commits/594716abdc84473dfcf3ced97dc9fb5897d5666c

emilyrohrbough commented on packages/terra-form/README.md:
```diff
@@ -0,0 +1,25 @@
+[![NPM version](http://img.shields.io/npm/v/terra-form.svg)](https://www.npmjs.org/package/terra-form)
+[![Build Status](https://travis-ci.org/cerner/terra-core.svg?branch=master)](https://travis-ci.org/cerner/terra-core)
+
+{insert description}
```
Add description

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/29d827a324b359afdb14063424ae363f20a44d3f

emilyrohrbough commented on packages/terra-form/docs/themeable-variables.md:
```diff
@@ -0,0 +1,14 @@
+| Variable Name                  | Default value            | Usage                                     |
+|--------------------------------|--------------------------|-------------------------------------------|
+| $terra-form-required-color     | $terra-red-70            | Color used to denote a field as required. |
+| $terra-form-required-indicator | *                        | Indicator to use for required fields.     |
```
Is the $terra-form-required-indicator a themeable string value?

emilyrohrbough commented on packages/terra-form/lib/_variables.scss:
```diff
@@ -0,0 +1,8 @@
+$terra-form-legend-font-weight: bold;
+$terra-form-field-border: 0;
+$terra-form-legend-color: $terra-dark;
+$terra-form-input-border: $terra-grey-15 solid 1px;
```
@neilpfeiffer Did we decided if themeable variables like border is themeable by color or also by border width, etc?

JakeLaCombe replied:
I didn't see form theming requirements, so I removed them.

https://github.com/cerner/terra-core/pull/352/commits/0943e9d4159fa1dd3f2e8325ede53d48c666d717

JakeLaCombe replied:
Removed, as I couldn't find theming requirements. https://github.com/cerner/terra-core/pull/352/commits/12196782a827e85c45446500ecbf1c7eef08b4e6

emilyrohrbough commented on packages/terra-form/package.json:
```diff
@@ -0,0 +1,57 @@
+  "main": "lib/Form.js",
+  "private": true,
+  "version": "0.0.0",
+  "description": "terra-form",
```
Add description

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/6242595b5b52cfa5f695be1f0e540a009c6f40de

emilyrohrbough commented on packages/terra-form/package.json:
```diff
@@ -0,0 +1,57 @@
+  },
+  "scripts": {
+    "compile": "npm run compile:clean && npm run compile:build",
+    "compile:clean": "rm -rf lib",
```
Update to use rimraf 
`"compile:clean": "$(cd ..; npm bin)/rimraf lib",`

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/0aee37909843eb8fc0b95a20a9f2868c48763be3

emilyrohrbough commented on packages/terra-form/src/components/Field.jsx:
```diff
@@ -0,0 +1,66 @@
+      {children}
+      {help && <small className="terra-Form-helpText">{help}</small>}
+      {error && <small className="terra-Form-error" tabIndex="-1">{error}</small>}
+    </p>
```
suggestion from old PR: https://github.com/cerner/terra-core/pull/305/files#diff-89873514d1b4737a5dd30a8c352d0088R53

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/68361e531d04900062a343d7b777983c79df4334

emilyrohrbough commented on packages/terra-form/src/components/Fieldset.jsx:
```diff
@@ -0,0 +1,65 @@
+
+const Fieldset = ({ children, error, help, htmlFor, isInline, legend, ...customProps }) => {
+  const fieldsetClasses = classNames(
+    'terra-Form-field',
```
Would we want a specific css class for fieldset to indicate the difference between a field component and a fieldset component?

JakeLaCombe replied:
Right now in blue_steel, we keep them using the same. I debated on this for awhile with myself. http://snapshot.bluesteel.hax/forms/layouts

However, I think for future maintainability, it's better to have them separate

JakeLaCombe replied:
Added the new class. https://github.com/cerner/terra-core/pull/352/commits/3474f23aa78e34bde9b8047d43f69b2696e8ce07

emilyrohrbough commented on packages/terra-form/tests/jest/components/Input.test.jsx:
```diff
@@ -0,0 +1,20 @@
+
+  const wrapper = shallow(input);
+  expect(wrapper).toMatchSnapshot();
+});
```
Is there value in adding snapshot tests for the differences between controlled and uncontrolled components?

JakeLaCombe replied:
@emilyrohrbough See how these work for you. From looking at the jest docs, I think this is the best I can get, which I think is ok for our case.

https://github.com/cerner/terra-core/pull/352/commits/f57bf091b3c5985efe503336c3b5fb4c5ae50896

emilyrohrbough replied:
@JakeLaCombe Looks good. Did the snapshot file update form this?

JakeLaCombe replied:
For this one, it didn't, as we didn't have new snapshots to check against.

emilyrohrbough commented on packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js:
```diff
@@ -0,0 +1,29 @@
+      .assert.elementPresent('.terra-Form-field--inline')
+      .assert.containsText('.terra-Form-legend', 'Do you have any Children?')
+      .assert.containsText('.terra-Form-helpText', 'Families are eligible for family package plans')
+      .assert.containsText('.terra-Form-error', 'This field is required');
```
How are these error messages being displayed?

JakeLaCombe replied:
![screen shot 2017-05-16 at 1 21 11 pm](https://cloud.githubusercontent.com/assets/2552573/26121717/d0339950-3a3a-11e7-8755-4a13d1b42817.png)

We have fieldset errors looking like this atm.

emilyrohrbough replied:
Okay. Wasn't sure if the error message was rendered after the user supplied incorrect input.

emilyrohrbough commented on packages/terra-site/src/examples/form/examples/Input.jsx:
```diff
@@ -0,0 +1,13 @@
+  </form>
+);
+
+export default InputExamples;
```
Should we include both a controlled & uncontrolled example?

JakeLaCombe replied:
My plan is to add a form example at the bottom of the page to handle that for all components. How does that sound?

JakeLaCombe replied:
I decided to add it for this pull request. Setups the easy addition of future controlled inputs.

https://github.com/cerner/terra-core/pull/352/commits/33ffe7b0aba15bcf8a600b929fea2d7f7d0c3e30

emilyrohrbough commented:
Missing jest tests for Field and Fieldset.

JakeLaCombe commented:
Wow! I had them on another branch, and it looks like I forgot to copy them over. https://github.com/cerner/terra-core/pull/352/commits/9add5acb750961b3d5acda03fe56aa47534644e8

Matt-Butler commented on packages/terra-form/src/components/Input.jsx:
```diff
@@ -0,0 +1,79 @@
+const defaultProps = {
+  defaultValue: undefined,
+  name: null,
+  onChange: () => {},
```
I think this should be removed.  If onChange is not provided, then no onChange function should be added to the input.

JakeLaCombe replied:
I agree. The easiest way to do this is to remove onChange as a first class attribute from the component, as it comes default.

https://github.com/cerner/terra-core/pull/352/commits/41cbef501d514b9499cdb298b1576db3bdd7195a

Matt-Butler commented on packages/terra-site/package.json:
```diff
@@ -41,6 +41,7 @@
     "terra-content-container": "^0.x",
     "terra-date-picker": "^0.x",
     "terra-demographics-banner": "^0.x",
+    "terra-form": "^0.x.0",
```
```
"^0.x"
```

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/2e129c61e6972077c23e0053be803b11b886b0ca

Matt-Butler commented:
I am seeing this warning
```
Warning: Input is changing a controlled input of type text to be uncontrolled. Input elements should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
```

![controlled_uncontrolled](https://cloud.githubusercontent.com/assets/8146159/26125191/e15607d4-3a46-11e7-9346-aabaefa0d79c.gif)

Matt-Butler commented:
![focus](https://cloud.githubusercontent.com/assets/8146159/26125218/f7286db8-3a46-11e7-9549-ee9d62290599.gif)
Should this element be focusable?

Matt-Butler commented on packages/terra-form/src/components/Input.jsx:
```diff
@@ -0,0 +1,79 @@
+    additionalInputProps['aria-required'] = 'true';
+  }
+
+  if (value) {
```
`""` an empty string will evaluate to false.

https://developer.mozilla.org/en-US/docs/Glossary/Falsy

Matt-Butler replied:
I would update this to check if value is undefined.

JakeLaCombe replied:
That did the trick.

https://github.com/cerner/terra-core/pull/352/commits/9c897f626b62fba1f2b35c6d5d000b7d911d6030

windse7en commented:
The heroku reviews app can't be created due to the existing app name: terra-core-deployed-pr-352.
I tried to access to that server, but I don't have the access. Someone created a heroku app with our pr-352 name.

To deploy to the heroku, my suggest:
1. We can create a heroku app from your form branch, normally deploy as heroku app.
2. Recreate another PR to avoid this name conflict.

JakeLaCombe commented:
New deploy url! http://terra-forms.herokuapp.com/static/#/site/form

JakeLaCombe commented:
@Matt-Butler The error messages have an interesting story behind them. I looked at the existing terra base, and found that all of the error and help messages had tabindex=-1 on them. This is due to issues with IE screen readers not being able to properly handle the aria-labelledby and aria-describedby attributes in the MSAA accessibility API. To fix this, microsoft provided a fix of attaching tabindex=-1 to the desired nodes. You can read more about it [here](https://www.paciellogroup.com/blog/2014/06/aria-labelledby-aria-describedby-support-popular-windows-browsers/).

To fix what you saw, I removed the outline from the error and help messages, and added tabindex=-1 to the help messages. https://github.com/cerner/terra-core/pull/352/commits/970e6a5f5b53c0672eec8f85511c5dc5f6cad3a1

windse7en commented on packages/terra-form/docs/components/Field.md:
````diff
@@ -0,0 +1,23 @@
+
+```jsx
+import React from 'react';
+import Field from 'terra-form/lib/Fieldset';
````
Is there any way we can get rid the `lib`,  just use `import Field from 'terra-form/Fieldset'`?

JakeLaCombe replied:
I envisioned having the following file structure for terra-form.

     - terra-form
        - lib
              - components
              - validations

It gives the component better file structure in my opinion. Do others have any views on this?

mjhenkes replied:
Do you anticipate having any additional folders at the lib level? I lean towards agreeing with @windse7en

JakeLaCombe replied:
Those are the only ones I anticipate atm. Given that, I can move everything up to lib.

JakeLaCombe replied:
Moved them over. https://github.com/cerner/terra-core/pull/352/commits/70bc8f0a6b07a540ff52743bd71d7063fba8fe5f

windse7en commented on packages/terra-form/src/components/Field.jsx:
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import 'terra-base/lib/baseStyles';
```
A stupid question: Why do our components have to import `baseStyles`?

mjhenkes replied:
@bjankord has a good explanation here: https://github.com/cerner/terra-core/pull/251

windse7en replied:
Thanks.

mjhenkes commented on packages/terra-form/src/components/Field.jsx:
```diff
@@ -0,0 +1,66 @@
+   */
+  help: PropTypes.node,
+  /**
+   * Determines whether the field in an inline field
```
"Determines whether the field is an inline field"

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/d43575884989c3f9802166d288225744244ad996

mjhenkes commented on packages/terra-form/src/components/Fieldset.jsx:
```diff
@@ -0,0 +1,71 @@
+   */
+  htmlFor: PropTypes.string,
+  /**
+   * Determines whether the field in an inline field
```
"Determines whether the field is an inline field"

JakeLaCombe replied:
https://github.com/cerner/terra-core/pull/352/commits/d43575884989c3f9802166d288225744244ad996

bjankord commented on packages/terra-form/docs/Fieldset.md:
```diff
@@ -0,0 +1,23 @@
+import Fieldset from 'terra-form/lib/Fieldset';
+
+<Fieldset
+  type="checkbox"
```
Is type a valid prop here? I didn't see it defined in the proptypes for `packages/terra-form/src/Fieldset.jsx`.

JakeLaCombe replied:
This is outdated. Fixed here. https://github.com/cerner/terra-core/pull/352/commits/33ffe7b0aba15bcf8a600b929fea2d7f7d0c3e30

bjankord commented on packages/terra-form/docs/Fieldset.md:
```diff
@@ -0,0 +1,23 @@
+  type="checkbox"
+  legend="Do you have any Children?"
+  name="children_present"
+  value="children_present"
```
Are `name` and `value` valid props here? I didn't see it defined in the proptypes for `packages/terra-form/src/Fieldset.jsx`.

JakeLaCombe replied:
Looks like I missed a few revisions from before. 

https://github.com/cerner/terra-core/pull/352/commits/33ffe7b0aba15bcf8a600b929fea2d7f7d0c3e30

nickpith commented on packages/terra-form/src/Fieldset.jsx:
```diff
@@ -0,0 +1,71 @@
+  required: false,
+};
+
+const Fieldset = ({ children, error, help, htmlFor, isInline, legend, required, ...customProps }) => {
```
I don't see htmlFor being used anywhere in this component.

JakeLaCombe replied:
It doesn't belong here. Removed.

https://github.com/cerner/terra-core/pull/352/commits/2c7f7c6030aea22430c5009637aca5681b94a9a6

JakeLaCombe commented:
Added htmlFor to the field. https://github.com/cerner/terra-core/pull/352/commits/c7029abf4f24439a2f1691787ab33ecd471e8766

I can't find a good way to generate unique ids for the htmlFor attribute, so I left that for the consumer to pass in. It's not throwing warnings in the browser or lint though, so it should be ok to leave the default as undefined.

mmalaker commented:
+1 functional review

neilpfeiffer commented:
+1 for checkpoint UX Review

One thing we will want to look at is the how to better style fieldsets, particularly the helptext and errors which look a little distant when fields are inline and don't also have helptext or errors. I'll make a note of some variations we can try.
